v2.0 Add unit tests for `as_gt`

edelarua opened the pull request:
**What changes are proposed in this pull request?**

Add v2.0 unit tests for `as_gt`.

Part of https://github.com/insightsengineering/sme-tasks/issues/689

Closes https://github.com/insightsengineering/sme-tasks/issues/692

--------------------------------------------------------------------------------

Reviewer Checklist (if item does not apply, mark is as complete)

- [ ] Ensure all package dependencies are installed by running `renv::install()`
- [ ] PR branch has pulled the most recent updates from master branch. Ensure the pull request branch and your local version match and both have the latest updates from the master branch.
- [ ] If an update was made to `tbl_summary()`, was the same change implemented for `tbl_svysummary()`?
- [ ] If a new function was added, function included in `_pkgdown.yml`
- [ ] If a bug was fixed, a unit test was added for the bug check
- [ ] Run `pkgdown::build_site()`. Check the R console for errors, and review the rendered website.
- [ ] Code coverage is suitable for any new functions/features. Review coverage with `withr::with_envvar(new = c("NOT_CRAN" = "true"), covr::report())`. Begin in a fresh R session without any packages loaded. 
- [ ] R CMD Check runs without errors, warnings, and notes
- [ ] `usethis::use_spell_check()` runs with no spelling errors in documentation

When the branch is ready to be merged into master:
- [ ] Update `NEWS.md` with the changes from this pull request under the heading "`# gtsummary (development version)`". If there is an issue associated with the pull request, reference it in parentheses at the end update (see `NEWS.md` for examples).
- [ ] Increment the version number using `usethis::use_version(which = "dev")` 
- [ ] Run `codemetar::write_codemeta()`
- [ ] Run `usethis::use_spell_check()` again
- [ ] Approve Pull Request
- [ ] Merge the PR. Please use "Squash and merge".

edelarua commented:
@ddsjoberg are these types of tests sufficient for this kind of function? I just went through the elements of the `gtsummary` objects and checked them against the equivalent attribute of the `gt` object. The old version used snapshot tests of the `gt` tables converted to html code, which I can add back in as well if you'd like.

ddsjoberg commented on tests/testthat/test-as_gt.R:
```diff
@@ -0,0 +1,290 @@
+    my_tbl_summary$table_styling$footnote$column,
+    gt_tbl_summary$`_footnotes`$colname |> unique()
+  )
+  expect_equal(
+    my_tbl_summary$table_styling$footnote$footnote,
+    gt_tbl_summary$`_footnotes`$footnotes |> unlist() |> unique()
+  )
```
Can we also test the scenario where a user replaces the default footnotes?

edelarua replied:
Added!

ddsjoberg commented on tests/testthat/test-as_gt.R:
```diff
@@ -0,0 +1,290 @@
+  )
+})
+
+test_that("as_gt passes table indentation correctly", {
```
Can we add a test where indentation has been removed?

edelarua replied:
Added!

ddsjoberg reviewed (changes requested):
@edelarua ! This is great!! You know a gt_tbl object so well! I think a couple of small changes and this is ready for review!

- Can we add tests for bold and italic cells?
- Tests that formatted missing values are correctly rendered.
- Tests the formatting functions are applied correctly (both to an entire column and to a cell)?
- Lastly, just a single test of column merging. When I wrote this originally, gt couldn't merge columns, but now they can. So I want to update this to use the gt merging tool. But also want a single test using the current structure so we don't forget to update/add more tests once we get around to making the merge update.

edelarua commented on tests/testthat/test-as_gt.R:
```diff
@@ -0,0 +1,290 @@
+  tbl_summary(by = grade, include = age) |>
+  modify_spanning_header(c(stat_1, stat_3) ~ "**Testing**")
+
+expect_silent(gt_tbl_summary <- my_tbl_summary |> as_gt())
+expect_silent(gt_tbl_cross <- my_tbl_cross |> as_gt())
+expect_silent(gt_tbl_regression <- my_tbl_regression |> as_gt())
+expect_silent(gt_tbl_uvregression <- my_tbl_uvregression |> as_gt())
+expect_silent(gt_spanning_tbl <- my_spanning_tbl |> as_gt())
```
If we want to reuse some of these and reduce duplication they can't be wrapped in a `test_that` chunk which stores all variables locally. I moved some of them which were not reused to inside their respective chunks, but what do you want to do about the others (e.g. `gt_tbl_summary` which is used in 4 tests)? I could just remove these lines and apply `as_gt` within each test as needed since I think `as_gt` is pretty quick... I have no strong opinion either way though I don't prefer having tests outside of a `test_that` chunk. Let me know what you think!

ddsjoberg replied:
Ahhh that makes sense. I don't have a strong opinion on it either. I guess my preference is to not wrap them in `expect_*()` fns at all. If they do fail, the testing procedure would fail too

edelarua commented on tests/testthat/test-as_gt.R:
```diff
@@ -0,0 +1,290 @@
+expect_silent(gt_spanning_tbl <- my_spanning_tbl |> as_gt())
+
+test_that("as_gt works with standard use", {
+  expect_silent(my_tbl_summary |> as_gt(return_calls = TRUE))
```
Added!

edelarua commented on tests/testthat/test-as_gt.R:
```diff
@@ -0,0 +1,290 @@
+  )
+
+  # tbl_uvregression
+  expect_equal(
+    my_tbl_uvregression$table_body |>
+      dplyr::select(-conf.low),
+    gt_tbl_uvregression$`_data` |>
+      dplyr::select(-conf.low),
+    ignore_attr = "class"
+  )
```
Since `conf.low` is unformatted in the `gtsummary` `table_body` and all formatting is applied within the `gt` `_data` object these two columns don't match.  Is there a function like `apply_fmt_fn` in `cards` that applies formatting functions and/or column merging formatting to gtsummary `table_body` objects? If so that would also be helpful for adding tests for formatting functions since I don't want to overcomplicate if possible.

ddsjoberg replied:
In this case, could you use `as_tibble()` or `as.data.frame()`? Those apply all the formatting functions, merging, etc. If you need more columns in `.$table_body` for testing, you can can run `modify_column_unhide()` and they'll be kept when converted to a tibble.

edelarua commented on tests/testthat/test-as_gt.R:
```diff
@@ -0,0 +1,290 @@
+  )
+
+  # customize interpreter
+  tbl <- my_spanning_tbl |>
+    modify_table_styling(columns = "stat_1", label = "Stat I", spanning_header = "MyTest", text_interpret = "html")
+  gt_tbl <- tbl |> as_gt()
+  if ("html" %in% names(attributes(gt_tbl$`_spanners`$spanner_label[[1]]))) {
+    attr(gt_tbl$`_spanners`$spanner_label[[1]], "html") <- c("html", "character")
+  }
```
This test has been simplified and this section was removed.

edelarua commented:
Hi @ddsjoberg,

I have added all but the formatting functions test (I made a comment regarding this test [here](https://github.com/ddsjoberg/gtsummary/pull/1727#discussion_r1643674133)) that you mentioned above and addressed the other code review comments.  

One additional question - currently the test for replacing a footnote with a new one is failing since the footnote that was removed is retained in the `table_styling$footnote` tibble. Is this the expected behaviour and if so, how should we compare these footnotes to the `_footnotes` tibble of the `gt` object which removes this information? Let me know how you'd like to proceed. 

Thanks!!

ddsjoberg commented:
> One additional question - currently the tests for replacing a footnote with a new one is failing since the footnote that was removed is retained in the table_styling$footnote tibble. Is this the expected behaviour and if so, how should we compare these footnotes to the _footnotes tibble of the gt object which removes this information? Let me know how you'd like to proceed.

That is indeed the expected behavior from the gtsummary side. We can just add a test that the new footnote appears in the gt table, and don't need to do an exact comparison to the gtsummary object...we can test against a hard-coded vector that is the footnote?

edelarua commented:
@ddsjoberg ready for another review!

ddsjoberg reviewed (approved):
@edelarua this is wonderful work. thank you so much for the effort!! SO MUCH APPRECIATED!!
